### PR TITLE
Refactor fault injection logic

### LINKED ISSUE #301: Come to better abstractions for fault injection
Presently, fault injection capabilitites are defined in a series of interfaces that defines a method for each type of fault. 
For example, the `ProtocolFaultInjector` defines the methods for injecting HTTP and gRPC faults. 

Each disruptor (PodDisruptor, ServiceDistruptor) has to implement these methods. 

The implementation of each fault injection method for each disruptor is quite similar, differing mostly in the type of fault to be injected. To inject a fault the disruptor needs to translate this fault to a command to be sent to the agent in the target using an agent controller.

This situation creates a series of issues:
- Duplicity of logic in the fault injection between disruptors and between protocols. This causes inconsistencies such as the one reported in #299 
- Complexity of tests: presently to validate how a fault is mapped to the corresponding agent commands it is necessary to instantiate the disruptors. This requires considerable setup and also creates the necessity to duplicate the tests for each disruptor
- No clear separation of concerns

All the above seems to point to the need to reconsider the abstractions used for fault injection. In particular to decouple the fault injection from the disruptor to allow reuse across disruptors. 

One initial analysis points to the idea of separating the fault injection logic into a component that has the sole responsibility of mapping faults to commands for a given target, validating that this mapping is possible/allowed.

### PR DESCRIPTION
**Pull request by pablochacin**
# Description

The purpose of the PR is to introduce abstractions that help in the separation of concerns around fault injection with the purpose of facilitating testing and promoting reusability.

Presently, the fault injection is done in conjunction between the disruptors (Pod, Service) and the AgentController. 

The AgentController encapsulates the logic for interacting with the agents in a set of targets. The disruptors encapsulate the logic for defining the list of targets and for creating the agent command for injecting a particular fault in a target.

However, the way these aspects are presently integrated difficult testing them separately. In particular, testing the generation of commands for fault injection requires setting a controller.

This problem comes primarily due to the way the interface between the controller and the disruptor is implemented. The disruptor passes an anonymous function to the controller that generates the commands for each target pod.  The logic of these functions cannot be tested without setting a controller which will initiate the call to the functions.

It is important to keep in mind that this model of interacting between the disruptors and the controller (the disruptor calls the controller's visit function that calls back the disruptor) was introduced to allow a per-target validation and customization of the commands. 

This PR introduces a new interface for generating these commands and the implementations for each type of fault. In this way, 
the command generation can be implemented independently of the controller.

It is important to notice that in its current form, this is a **naive implementation with learning purposes**, to help in identifying common patterns that could be refactored into existing or new abstractions. 

For example, the fault injection for each fault and target type is implemented as a different struct. Subsequent changes could introduce alternatives.

Fixes #301

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works.   
- [ ] I have run linter locally (`make lint`) and all checks pass.
- [x] I have run tests locally (`make test`) and all tests pass.
- [x] I have run relevant integration tests locally 
- [x] I have run relevant e2e tests locally 
- [ ] Any dependent changes have been merged and published in downstream modules<br>

### CODE REVIEW

File: pkg/disruptors/injector.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	return visitCommands, nil
+}
+
+// ServiceHTTPFaultInjector implements the Visitor interface for injecting HttpFaults in a Pod
+type ServiceHTTPFaultInjector struct {
+	service  corev1.Service
+	fault    HTTPFault
+	duration time.Duration
+	options  HTTPDisruptionOptions
+}
```
Comment (roobre) [inline pkg/disruptors/injector.go:73]: Should we add tests for `ServiceHTTPFaultInjector` and/or `ServiceGrpcFaultInjector`? If we have some already, I'm not being able to find them 🤔

Comment (pablochacin) [reply pkg/disruptors/injector.go:73]: I think that's probably out of scope here. I would prefer a follow-up PR doing this and closing https://github.com/grafana/xk6-disruptor/issues/261

File: pkg/disruptors/injector_test.go
[the remark this answers is not in the document]
Comment (pablochacin) [reply pkg/disruptors/injector_test.go:1]: This is probably unrelated to the purpose of this PR, so I prefer to leave it to another PR.

[the remark this answers is not in the document]
Comment (pablochacin) [reply pkg/disruptors/injector_test.go:1]: I actually renames to `Test_PodxxxFaultVisitor` following the new naming suggested [here](https://github.com/grafana/xk6-disruptor/pull/307#discussion_r1331284718)

**Review by roobre — APPROVED** [review]
LGTM!

**Pull request merged**